### DISCUSSION TITLE
Discussion about multipart post with Async::HTTP::Client/Falcon

### DISCUSSION THREAD

**Discussion by renatolond** [socketry/async-http#193] [General]
Hey!

I was checking https://github.com/socketry/async-http/discussions/160 about multipart post, but I have a slightly different case and I feel like what I'm doing is a bit odd.

I have a rails app (running Falcon) with a form to post an image. The rails app then takes that image (that comes in an ActionDispatch::Http::UploadedFile) and will forward it to a backend using a Async::HTTP::Client.

The code in the controller looks something like this:
```
client.upload_profile_picture(filename: params[:profile_picture].original_filename, io: params[:profile_picture].to_io, content_type: params[:profile_picture].content_type)
```

And in the client object:
```
  # @param filename [String] The name of the file
  # @param io [File] The file to be uploaded
  # @param content_type [String] The content type of the original file
  # @return [void]
  def upload_profile_picture(filename:, io:, content_type:)
    return nil if @authorization_header.blank?

    form_data = {
      file: ::Multipart::Post::UploadIO.new(io, content_type, filename)
    }
    headers = {}
    multipart_post = ::Net::HTTP::Post::Multipart.new("/api/profile/picture", form_data, headers)
    headers["Content-Type"] = multipart_post["content-type"]
    Sync do
      response = async_client.post("/api/profile/picture", headers: base_headers.merge(headers), body: multipart_post.body_stream.read)
      case response.status
      when 204
        puts "Success!"
      end
    ensure
      response&.close
    end
  end
```

I used part of the discussion in #160 to make this and it works (full code over at https://github.com/retromeet/web-client/commit/0f4d33f0b164102ce5d94537afd7ac46a83ea1f1). Besides the oddness of this "manual redirection" of sorts, it feels odd to load part of Net::HTTP for this to work.

From your explanation, I believe a `Protocol::HTTP::Multipart` class at the very leat would have to derive from `Readable` and re-implement the logic of `Multipart::Post::Multipartable`, does it make sense?

**Comment by ioquatix** [comment]
Yes, that makes sense, I should probably work on `protocol-multipart`.
